// @confio/relayer/build/main/binary/ibc-setup/commands/balances.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const os_1 = __importDefault(require("os"));
const ava_1 = __importDefault(require("ava"));
const sinon_1 = __importDefault(require("sinon"));
const ibcclient_1 = require("../../../lib/ibcclient");
const testutils_1 = require("../../../lib/testutils");
const balances_1 = require("./balances");
const fsReadFileSync = sinon_1.default.stub(fs_1.default, 'readFileSync');
const mnemonic = 'accident harvest weasel surge source return tag supreme sorry isolate wave mammal';
function buildIbcArgs(rpc) {
    return [rpc, sinon_1.default.match.any, sinon_1.default.match.any, sinon_1.default.match.any];
}
const ibcClient = sinon_1.default.stub(ibcclient_1.IbcClient, 'connectWithSigner');
const musselnetArgs = buildIbcArgs('https://rpc.musselnet.cosmwasm.com:443');
const localWasmArgs = buildIbcArgs('http://localhost:26659');
const localSimappArgs = buildIbcArgs('http://localhost:26658');
async function createFakeIbcClient(amount, denom) {
    return {
        query: {
            bank: {
                unverified: {
                    balance: sinon_1.default.fake.returns({ amount, denom }),
                },
            },
        },
    };
}
ava_1.default.beforeEach(() => {
    sinon_1.default.reset();
});
const registryYaml = `
version: 1

chains:
  musselnet:
    chain_id: musselnet-4
    # bech32 prefix for addresses
    prefix: wasm
    # this determines the gas payments we make (and defines the fee token)
    gas_price: 0.025umayo
    # the path we use to derive the private key from the mnemonic
    hd_path: m/44'/108'/0'/1'
    # you can include multiple RPC endpoints and it will rotate through them if
    # one is down
    rpc:
      - https://rpc.musselnet.cosmwasm.com:443
  local_wasm:
    chain_id: testing
    prefix: wasm
    gas_price: 0.025ucosm
    hd_path: m/44'/108'/0'/2'
    rpc:
      - http://localhost:26659
  local_simapp:
    chain_id: simd-testing
    prefix: cosmos
    gas_price: 0.025umuon
    hd_path: m/44'/108'/0'/3'
    rpc:
      - http://localhost:26658`;
ava_1.default('lists chains with non-zero balance', async (t) => {
    const logger = new testutils_1.TestLogger();
    const options = {
        home: '/home/user',
        mnemonic,
    };
    fsReadFileSync.returns(registryYaml);
    ibcClient
        .withArgs(...musselnetArgs)
        .returns(createFakeIbcClient('1', 'musselnetdenom'))
        .withArgs(...localWasmArgs)
        .returns(createFakeIbcClient('2', 'wasmdenom'))
        .withArgs(...localSimappArgs)
        .returns(createFakeIbcClient('3', 'simappdenom'));
    await balances_1.run(options, logger);
    t.assert(fsReadFileSync.calledOnce);
    t.assert(logger.info.calledOnce);
    t.assert(logger.info.calledWithExactly([
        'musselnet: 1musselnetdenom',
        'local_wasm: 2wasmdenom',
        'local_simapp: 3simappdenom',
    ].join(os_1.default.EOL)));
});
ava_1.default('omits chains with zero balance', async (t) => {
    const logger = new testutils_1.TestLogger();
    const options = {
        home: '/home/user',
        mnemonic,
    };
    fsReadFileSync.returns(registryYaml);
    ibcClient
        .withArgs(...musselnetArgs)
        .returns(createFakeIbcClient('1', 'musselnetdenom'))
        .withArgs(...localWasmArgs)
        .returns(createFakeIbcClient('0', 'wasmdenom'))
        .withArgs(...localSimappArgs)
        .returns(createFakeIbcClient('3', 'simappdenom'));
    await balances_1.run(options, logger);
    t.assert(fsReadFileSync.calledOnce);
    t.assert(logger.info.calledOnce);
    t.assert(logger.info.calledWithExactly(['musselnet: 1musselnetdenom', 'local_simapp: 3simappdenom'].join(os_1.default.EOL)));
});
ava_1.default('informs when there are no funds on any balance', async (t) => {
    const logger = new testutils_1.TestLogger();
    const options = {
        home: '/home/user',
        mnemonic,
    };
    fsReadFileSync.returns(registryYaml);
    ibcClient
        .withArgs(...musselnetArgs)
        .returns(createFakeIbcClient('0', 'musselnetdenom'))
        .withArgs(...localWasmArgs)
        .returns(createFakeIbcClient('0', 'wasmdenom'))
        .withArgs(...localSimappArgs)
        .returns(createFakeIbcClient('0', 'simappdenom'));
    await balances_1.run(options, logger);
    t.assert(fsReadFileSync.calledOnce);
    t.assert(logger.info.calledOnce);
    t.assert(logger.info.calledWithMatch(/No funds/));
});
