// @confio/relayer/build/main/binary/ibc-setup/commands/connect.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const utils_1 = require("@cosmjs/utils");
const ava_1 = __importDefault(require("ava"));
const sinon_1 = __importDefault(require("sinon"));
const testutils_1 = require("../../../lib/testutils");
const constants_1 = require("../../constants");
const signing_client_1 = require("../../utils/signing-client");
const chains_1 = require("./chains");
const connect_1 = require("./connect");
const fsWriteFileSync = sinon_1.default.stub(fs_1.default, 'writeFileSync');
const fsReadFileSync = sinon_1.default.stub(fs_1.default, 'readFileSync');
const mnemonic = 'enlist hip relief stomach skate base shallow young switch frequent cry park';
const registryYaml = `
version: 1

chains:
  local_wasm:
    chain_id: testing
    prefix: wasm
    gas_price: 0.025ucosm
    rpc:
      - http://localhost:26659
  local_simapp:
    chain_id: simd-testing
    prefix: cosmos
    gas_price: 0.025umuon
    rpc:
      - http://localhost:26658`;
const app = {
    src: 'local_wasm',
    dest: 'local_simapp',
};
ava_1.default.beforeEach(() => {
    sinon_1.default.reset();
});
ava_1.default.serial('connects two chains', async (t) => {
    var _a, _b;
    const logger = new testutils_1.TestLogger();
    const ibcClientSimapp = await signing_client_1.signingClient(chains_1.simappChain, mnemonic);
    const ibcClientWasm = await signing_client_1.signingClient(chains_1.wasmdChain, mnemonic);
    const allConnectionsWasm = await ibcClientWasm.query.ibc.connection.allConnections();
    const allConnectionsSimapp = await ibcClientSimapp.query.ibc.connection.allConnections();
    const options = {
        home: '/home/user',
        mnemonic,
        src: 'local_simapp',
        dest: 'local_wasm',
    };
    fsReadFileSync.returns(registryYaml);
    fsWriteFileSync.returns();
    await connect_1.run(options, app, logger);
    const args = fsWriteFileSync.getCall(0).args;
    const contentsRegexp = new RegExp(`src: local_wasm
dest: local_simapp
srcConnection: .+
destConnection: .+
`);
    t.assert(fsWriteFileSync.calledOnce);
    t.is(args[0], path_1.default.join(options.home, constants_1.appFile));
    t.regex(args[1], contentsRegexp);
    t.assert(logger.info.calledOnce);
    t.assert(logger.info.calledWithMatch(/Created connections/));
    const nextAllConnectionsWasm = await ibcClientWasm.query.ibc.connection.allConnections();
    const destConnectionIdMatch = /destConnection: (?<connection>.+)/.exec(args[1]);
    const destConnectionId = (_a = destConnectionIdMatch === null || destConnectionIdMatch === void 0 ? void 0 : destConnectionIdMatch.groups) === null || _a === void 0 ? void 0 : _a.connection;
    utils_1.assert(destConnectionId);
    const nextConnectionWasm = await ibcClientWasm.query.ibc.connection.connection(destConnectionId);
    const nextAllConnectionsSimapp = await ibcClientSimapp.query.ibc.connection.allConnections();
    const srcConnectionIdMatch = /srcConnection: (?<connection>.+)/.exec(args[1]);
    const srcConnectionId = (_b = srcConnectionIdMatch === null || srcConnectionIdMatch === void 0 ? void 0 : srcConnectionIdMatch.groups) === null || _b === void 0 ? void 0 : _b.connection;
    utils_1.assert(srcConnectionId);
    const nextConnectionSimapp = await ibcClientSimapp.query.ibc.connection.connection(srcConnectionId);
    t.is(nextAllConnectionsWasm.connections.length, allConnectionsWasm.connections.length + 1);
    t.is(nextAllConnectionsSimapp.connections.length, allConnectionsSimapp.connections.length + 1);
    t.assert(nextConnectionWasm.connection);
    t.assert(nextConnectionSimapp.connection);
});
