// @confio/relayer/build/main/binary/ibc-setup/commands/keys-list.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const ava_1 = __importDefault(require("ava"));
const sinon_1 = __importDefault(require("sinon"));
const testutils_1 = require("../../../lib/testutils");
const generate_mnemonic_1 = require("../../utils/generate-mnemonic");
const keys_list_1 = require("./keys-list");
const fsReadFileSync = sinon_1.default.stub(fs_1.default, 'readFileSync');
ava_1.default.beforeEach(() => {
    sinon_1.default.reset();
});
const registryYaml = `
version: 1

chains:
  local_wasm:
    chain_id: testing
    prefix: wasm
    gas_price: 0.025ucosm
    hd_path: m/44'/108'/0'/2'
    rpc:
      - http://localhost:26659
  local_simapp:
    chain_id: simd-testing
    prefix: cosmos
    gas_price: 0.025umuon
    hd_path: m/44'/108'/0'/3'
    rpc:
      - http://localhost:26658`;
ava_1.default('lists addresses for every chain in the registry', async (t) => {
    const logger = new testutils_1.TestLogger();
    const options = {
        home: '/home/user',
        mnemonic: generate_mnemonic_1.generateMnemonic(),
    };
    fsReadFileSync.returns(registryYaml);
    await keys_list_1.run(options, logger);
    t.assert(fsReadFileSync.calledOnce);
    t.assert(logger.info.calledOnce);
    t.assert(logger.info.calledWithMatch(/local_wasm: [a-z0-9]+\nlocal_simapp: [a-z0-9]+/));
});
